### ISSUE TITLE
Energy deposition in TOF mismatch in data & MC

### ISSUE THREAD

**Issue by sdobbs** [JeffersonLab/halld_sim#128]
This issue comes from Andrew Schick, who is using a TMVA classifier to separate electrons from pions in order to measure g p -> e+ e- (p) at very small t.  One of the input he has found useful is the dE in the TOF, and as seen below, the energy scale between data and MC seems to be off by a factor of 2.

I'm not sure if this is a problem with the simulation the TOF in HDGeant4, or that the attenuation length assumed for the simulations is incorrect.  Any thoughts?

![Electron dEdx in TOF](https://user-images.githubusercontent.com/1182058/81848654-7060af80-9523-11ea-8f22-8103d2b16511.png)
![Positron dEdx in TOF](https://user-images.githubusercontent.com/1182058/81848656-7060af80-9523-11ea-8632-a3c3b2f630af.png)

**Comment by rjones30** [comment]
Hello all,

Yes, the attenuation is included in the MC output. That definitely needs to
be done in simulation before hits are summed. Once the hits are summed and
the pileup is accounted for, the attenuation can then be corrected for (to
leading order) in mcsmear. It looks to me like that is what is needed,
exactly what mcsmear is intended for.

-Richard J.

**Comment by zihlmann** [comment]
There was never a gain calibration done for TOF as it was never 
considered necessary.
There are plans to do a better gain matching HV adjustments to equalize 
the signal amplitudes
in particular for the CPP experiment since they want to use the TOF in 
the trigger.

**Comment by sdobbs** [comment]
Right, gain calibration is one thing, but at least in data, there's some reasonable peak as opposed to in the MC.

Looking at the attenuation length factors being used, in hdgeant4 the value is 150. cm (from TOF/tof_params), but in the reconstruction, it is generally 470. cm (from TOF/attenuation_lengths).  Am I correct here or missing something?  Should these values be more consistent?

**Comment by rjones30** [comment]
Sean,

Yes, we could decide to get these attenuation lengths for simulation from
ccdb. Is that the wish?

-Richard

**Comment by zihlmann** [comment]
The reality is of course much more complicated. If you really want to 
describe the attenuation
in a realistic way you need at a minimum two factors, a long one 470 and 
a short one ~100.
When fitting the data you find that one single value will not work at 
all. The choice of 150
is because if you look only at hits around the central region +-50cm or 
so (where most of the
hits come from), then 150cm does a reasonable job.

**Comment by rjones30** [comment]
For me, the important thing is that the hits get delayed, attenuated, when
they are superimposed in the simulation. This way truth information is not
secretly imprinted in the mc stream that cannot in principle be there in
the real data. The hits can then be rescaled and reshifted in mcsmear. Just
tell me how you want the delay and attenuation modeled and I will implement
it. Mcmsear can then rescale and reshape to match the data.

-Richard

**Comment by sdobbs** [comment]
Right, I was hoping to get some insight from Beni on what the right approach would be to apply the attenuation correction in the simulation since I don't have a good intuition on this.  I can ask Andrew to do some more systematic studies to narrow this down.

**Comment by zihlmann** [comment]
here is an example of how the attenuation in a paddle looks like as seen by the PMT. The horzontal axis is the time difference between Left and right PMT in ns and as such directly related to distance. The data is corrected so that zero represents the center of the paddle. As you can see a single exponential will not fit all the data however it could fit the central region where most of the hits will be.  The vertical axis is the location of the landau peak (MPV) from a fit to the PMT response.
![p46_s1_plane1](https://user-images.githubusercontent.com/13365259/82345821-b73b2300-99c3-11ea-8fa7-113ef61c63cf.gif)

**Comment by rjones30** [comment]
What is the exponential decay constant of this fit, in cm?
-Richard

**Comment by zihlmann** [comment]
the slope is of order 0.05 which is when transforming from time to distance
I get 1/0.05*16cm/ns = 320cm    I assume here the effective speed of light
in the bar is around 16cm/ns

**Comment by rjones30** [comment]
Ok, the attenuation constant that is hard-coded into (both) hdgeant and
hdgeant4 is much too small, then.

-Richard

**Comment by zihlmann** [comment]
Using run 42559 and analyzing the first DATA file with the new code that also uses the proper energy calibration for the tof from ccdb shows the energy deposition in the TOF correctly at about 4.6MeV. The Monte Carlo simulation shows values that are a factor 2.3 lower and has to be addressed in the simulation. Note that the track was required to be at 5 degrees +/- 1degree, so radius with theta 5 degree.
Both plots below for data and MC have a fit with a landau function that describes the data rather well. Scaling the width according to the MPV shows that the MC seems to describe the DATA already reasonably well in terms of width. But this has to be verified.
I have not used TMVA but that initial plot does not make any sense as dEdx should be about 1.8MeV at the max since the TOF paddles are 1 inch thick and the path length variation through the paddles is not very large.

DATA
![r42559_DATA](https://user-images.githubusercontent.com/13365259/85281673-6bfbb080-b458-11ea-874b-b59118a0a73d.gif)

MC
![r42559_MC](https://user-images.githubusercontent.com/13365259/85281691-7027ce00-b458-11ea-884e-f6831df97e65.gif)

**Comment by rjones30** [comment]
Beni, I will have a look, but my first guess is that you have not properly
corrected for the attenuation in the MC.

-Richard J.

**Comment by sdobbs** [comment]
@zihlmann - this is great to see, just as a sanity check, I wonder if you've done a similar comparison for pions with the latest calibrations in data / MC?

**Comment by zihlmann** [comment]
yes the pions and leptons pretty much look the same in real data. I 
checked that also with particle gun MC
that also show that the lepton reconstruction suffers significantly due 
to radiation effects.

**Comment by rjones30** [comment]
The way the simulation delta-E in the TOF is being handled is as follows:

   1. element in hddm we are discussing is dE in <tofTruthHit dE="float">
   2. in the simulation, ftofTruthHit dE = actual_dE * exp(-dist /
   ATTEN_LENGTH) where dist is distance from the hit to the corresponding end
   of the paddle, north or south
   3. in mcsmear: smeared dE = titer->getDE()  + Poisson noise

The ATTEN_LENGTH is taken from ccdb table TOF/tof_parms/TOF_ATTEN_LENGTH
one entry per counter.

Looking at the real hits decoding in DTOFHit_factory.cc, I see that no
attenuation correction is being applied when TOF_dE is being created.
Therefore BOTH mc and real data SHOULD have attenuation built into the dE
spectrum. That is, your TOF_dE plot for real data should be looking like
the MC plot, and not vice versa. This is a matter of definition of what
TOF_dE is supposed to mean, not physics. Either it should be attenuation
corrected or not. If not (the present code does NOT) then the measured
spectra should peak somewhere around 2 MeV in both real data and MC.

-Richard Jones

**Comment by zihlmann** [comment]
1) The attenuation  correction is applied in DTOFPaddle_factory since 
you need the hit position along the paddle
    to determine the attenuation.
2) For the short paddles (single ended readout) this attenuation is 
applied in the DTOFPoint_factory where
    the hit position can be inferred from the paddle location in the 
other plane and an attenuation correction
    is applied.
3) the energy conversion from ADC counts to MeV is done very early in 
DTOFHit_factory where the hit position
    is obviously ignored as it is not available at that stage.
4) the attenuation is applied such that the reference point is the 
center of the long paddle which is zero
    in x for horizontal paddles and in y for vertical paddles. For the 
short paddles the reference point is
    chosen close the the end (beam line) of the paddle.
5) The energy deposition in a paddle by a minimum ionizing particle is 
about 4.6MeV since the paddle is one inch
    thick. And that is what the DATA will give you in the 
DTOFPopint->dE, dE1 and dE2 objects. dE is the mean of
    dE1 and dE2 as it was always and therefor was not touched but dE1 
and dE2 were added to give the contributions
    from the two planes separately.

**Comment by rjones30** [comment]
Beni,

Everything you say is consistent with what I said. So we agree?

-Richard J.

**Comment by zihlmann** [comment]
Ah I guess I see now the confusion.
I am not plotting dE of the TOF hit but dE of the TOF point.

**Comment by rjones30** [comment]
Ah ok, I will look into it.
-Richard

**Comment by rjones30** [comment]
I show below the output from hdgeant (green) and hdgeant4 (black) for a 10k sample of 1 GeV pi+ from a particle gun firing at a fixed angle of 5 degrees from the liquid hydrogen target into the forward region. Simulation pulse heights from Individual paddle-end hits in the two layers of the tof are shown below, after the simple exponential attenuation that is applied in the simulation has been removed. Good agreement is seen in the lineshape between both simulations, and they both agree with the expectation that the most probably value should be around 5 MeV. From this, I conclude that the problem is not in the simulations themselves.
![image](https://user-images.githubusercontent.com/7832920/86515564-c9dcb080-bde7-11ea-8bba-9860ac0f91e6.png)
![image](https://user-images.githubusercontent.com/7832920/86515569-d2cd8200-bde7-11ea-984a-f4d8df609946.png)
![image](https://user-images.githubusercontent.com/7832920/86515572-dcef8080-bde7-11ea-9621-05361f18a266.png)
![image](https://user-images.githubusercontent.com/7832920/86515576-e4168e80-bde7-11ea-944b-8fc393c1bf97.png)

**Comment by rjones30** [comment]
It is interesting to note, however the huge amount of low-energy noise that hdgeant4 puts in the tof paddles, which is only partially present in the hdgeant simulation. Most of these noise hits in hdgeant4 come much later in time than the prompt signal, as can be seen in the comparison of the time of these hits shown below. I only show one plane/end combination this time because to my eye they all look the same.
![image](https://user-images.githubusercontent.com/7832920/86515693-da415b00-bde8-11ea-8579-6850d48b8ca0.png)

**Comment by rjones30** [comment]
Here is a plot of the particle types that are causing these noise hits in the tof. To select the noise, I only look at hits later than 35ns in time. Here is a partial lookup table of the particle types: 1=gamma, 2=positron, 3=electron, 5=mu+, 6=mu-, 8=pi+, 9=pi-, 13=neutron, 14=proton, 45=deutron, 47=alpha. You can see that most of these are neutrons, as you might expect. Notice that hdgeant4 has a default energy cut at 10MeV for neutrons, so it does not even attempt to simulate the neutron flux. It has a similar 10MeV cutoff for protons, which is probably not bad in the case of plastic.
![image](https://user-images.githubusercontent.com/7832920/86515761-65225580-bde9-11ea-92ed-3a9a08c62ee8.png)
Here is the same image on a log scale, so you can see that hdgeant also produces noise hits from the same types of particles, just fewer of them.
![image](https://user-images.githubusercontent.com/7832920/86515866-2e007400-bdea-11ea-98f2-33a24f226352.png)

**Comment by rjones30** [comment]
The problem with the energy scale for dE in the TOF in simulation was traced back to a difference between simulation and reconstruction in how the attenuation correction is defined. In the simulation, the reference point where the attenuation factor =1 is at the end of the bar nearest the tube. In the reconstruction, the reference point where attenuation factor=1 is the middle of the bar. The ratio between these is exp(half_bar_length / 150cm) = 2.3 which is the mismatch factor in the two energy scales we are seeing. Here are the results for DTOFPoint::dE with my fix to mcsmear applied.
![image](https://user-images.githubusercontent.com/7832920/86517390-98b7ac80-bdf6-11ea-954e-3cae64c8099c.png)
I have submitted a PR that implements this fix, based on the current simple attenuation law that is used by the two simulations.

Beni has suggested that a better description of the actual attenuation curves for the paddles should be used in the simulation. I am fine with that. Whatever we do will need to be implemented in a consistent fashion in both simulations and also at this spot in mcsmear where I applied this fix.
